File: fastapi/src/routes/attachment_routes.py
from fastapi import APIRouter, Depends

from src.schemas.attachment_schemas import (
    AttachmentCreate,
    AttachmentListResponse,
    AttachmentResponse,
    AttachmentUploadResponse,
)
from src.schemas.common_schemas import ErrorResponse
from src.services.attachment_service import AttachmentService, get_attachment_service

router = APIRouter(prefix="/tasks/{task_id}/attachments", tags=["Attachments"])


@router.get(
    "",
    response_model=AttachmentListResponse,
    summary="List attachments for a task",
)
async def list_attachments(
    task_id: str,
    limit: int = 20,
    offset: int = 0,
    service: AttachmentService = Depends(get_attachment_service),
) -> AttachmentListResponse:
    from uuid import UUID

    items, meta = await service.list(
        task_id=UUID(task_id),
        limit=limit,
        offset=offset,
    )
    return AttachmentListResponse(
        data=[
            AttachmentResponse(
                id=a.id,
                task_id=a.task_id,
                file_name=a.file_name,
                file_key=a.file_key,
                mime_type=a.mime_type,
                file_size=a.file_size,
                uploaded_by=a.uploaded_by,
                created_at=a.created_at,
            )
            for a in items
        ],
        meta=meta,
    )


@router.post(
    "",
    response_model=AttachmentUploadResponse,
    status_code=201,
    summary="Create attachment (returns presigned upload URL)",
)
async def create_attachment(
    task_id: str,
    params: AttachmentCreate,
    service: AttachmentService = Depends(get_attachment_service),
) -> AttachmentUploadResponse:
    from uuid import UUID

    attachment, upload_url = await service.create(
        task_id=UUID(task_id),
        params=params,
    )
    return AttachmentUploadResponse(
        upload_url=upload_url,
        attachment=AttachmentResponse(
            id=attachment.id,
            task_id=attachment.task_id,
            file_name=attachment.file_name,
            file_key=attachment.file_key,
            mime_type=attachment.mime_type,
            file_size=attachment.file_size,
            uploaded_by=attachment.uploaded_by,
            created_at=attachment.created_at,
        ),
    )


@router.get(
    "/{id}/download",
    response_model=dict[str, str],
    summary="Get presigned download URL",
    responses={404: {"model": ErrorResponse}},
)
async def get_download_url(
    id: str,
    service: AttachmentService = Depends(get_attachment_service),
) -> dict[str, str]:
    from uuid import UUID

    download_url = await service.get_download_url(UUID(id))
    return {"downloadUrl": download_url}

File: hono-node/src/routes/attachment-routes.ts
import { OpenAPIHono, createRoute, z } from "@hono/zod-openapi";

import type { HonoEnv } from "@/core/hono-env";
import {
  AttachmentCreateSchema,
  AttachmentListQuerySchema,
  AttachmentListResponseSchema,
  AttachmentUploadResponseSchema,
} from "@/schemas/attachment-schemas";
import { ErrorSchema } from "@/schemas/common-schemas";
import { attachmentService } from "@/services/attachment-service";

const attachmentRoutes = new OpenAPIHono<HonoEnv>();

const listAttachments = createRoute({
  method: "get",
  path: "/tasks/{taskId}/attachments",
  tags: ["Attachments"],
  summary: "List attachments for a task",
  request: {
    params: z.object({ taskId: z.string().uuid() }),
    query: AttachmentListQuerySchema,
  },
  responses: {
    200: {
      description: "List of attachments",
      content: {
        "application/json": { schema: AttachmentListResponseSchema },
      },
    },
  },
});

attachmentRoutes.openapi(listAttachments, async (c) => {
  const { taskId } = c.req.valid("param");
  const query = c.req.valid("query");
  const { items, meta } = await attachmentService.list({
    taskId,
    limit: query.limit,
    offset: query.offset,
  });
  return c.json(
    {
      data: items.map((a) => ({
        id: a.id,
        taskId: a.taskId,
        fileName: a.fileName,
        fileKey: a.fileKey,
        mimeType: a.mimeType,
        fileSize: a.fileSize,
        uploadedBy: a.uploadedBy,
        createdAt: a.createdAt.toISOString(),
      })),
      meta,
    },
    200,
  );
});

const createAttachment = createRoute({
  method: "post",
  path: "/tasks/{taskId}/attachments",
  tags: ["Attachments"],
  summary: "Create attachment (returns presigned upload URL)",
  request: {
    params: z.object({ taskId: z.string().uuid() }),
    body: {
      content: { "application/json": { schema: AttachmentCreateSchema } },
    },
  },
  responses: {
    201: {
      description: "Created with presigned URL",
      content: {
        "application/json": { schema: AttachmentUploadResponseSchema },
      },
    },
  },
});

attachmentRoutes.openapi(createAttachment, async (c) => {
  const { taskId } = c.req.valid("param");
  const body = c.req.valid("json");
  const { attachment, uploadUrl } = await attachmentService.create({
    taskId,
    fileName: body.fileName,
    mimeType: body.mimeType,
    fileSize: body.fileSize,
    uploadedBy: c.get("user")?.id,
  });
  return c.json(
    {
      uploadUrl,
      attachment: {
        id: attachment.id,
        taskId: attachment.taskId,
        fileName: attachment.fileName,
        fileKey: attachment.fileKey,
        mimeType: attachment.mimeType,
        fileSize: attachment.fileSize,
        uploadedBy: attachment.uploadedBy,
        createdAt: attachment.createdAt.toISOString(),
      },
    },
    201,
  );
});

const getDownloadUrl = createRoute({
  method: "get",
  path: "/attachments/{id}/download",
  tags: ["Attachments"],
  summary: "Get presigned download URL",
  request: { params: z.object({ id: z.string().uuid() }) },
  responses: {
    200: {
      description: "Download URL",
      content: {
        "application/json": {
          schema: z.object({ downloadUrl: z.string().url() }),
        },
      },
    },
    404: {
      description: "Not found",
      content: { "application/json": { schema: ErrorSchema } },
    },
  },
});

attachmentRoutes.openapi(getDownloadUrl, async (c) => {
  const { id } = c.req.valid("param");
  const downloadUrl = await attachmentService.getDownloadUrl(id);
  return c.json({ downloadUrl }, 200);
});

export default attachmentRoutes;
